Xamarin.Essentials/Types/FileProvider.android.cs

```diff
@@ -0,0 +1,102 @@
+﻿using System;
+using Android.App;
+using Android.Content;
+using Android.OS;
+using AndroidEnvironment = Android.OS.Environment;
+
+namespace Xamarin.Essentials
+{
+    [ContentProvider(
+        new[] { "${applicationId}.fileProvider" },
+        Name = "xamarin.essentials.fileProvider",
+        Exported = false,
+        GrantUriPermissions = true)]
+    [MetaData(
+        "android.support.FILE_PROVIDER_PATHS",
+        Resource = "@xml/xamarin_essentials_fileprovider_file_paths")]
+    public class FileProvider : global::Android.Support.V4.Content.FileProvider
+    {
+        internal static bool AlwaysFailExternalMediaAccess { get; set; } = false;
+
+        public static FileProviderLocation TemporaryLocation { get; set; }
+
+        internal static Java.IO.File GetTemporaryDirectory()
+        {
+            var root = GetTemporaryRootDirectory();
+            var dir = new Java.IO.File(root, "2203693cc04e0be7f4f024d5f9499e13");
+            dir.Mkdirs();
+            dir.DeleteOnExit();
+            return dir;
+        }
+
+        internal static Java.IO.File GetTemporaryRootDirectory()
+        {
+            // we want the internal storage
+            if (TemporaryLocation == FileProviderLocation.Internal)
+                return Platform.AppContext.CacheDir;
+
+            // if we need permission, then check
+            // if we can fall back, then don't throw
+            var externalOnly = TemporaryLocation == FileProviderLocation.External;
+            var hasPermission = Platform.HasApiLevel(BuildVersionCodes.Kitkat) || Permissions.EnsureDeclared(PermissionType.ExternalStorage, externalOnly);
```

Comment:
On 4.4+, we don't need to check for permissions to access the external cache or external app files. We do need to in order to access the downloads/documents, but we aren't doing that here.